### DISCUSSION TITLE
Auto Dark Mode X 10.3.0

### DISCUSSION THREAD

**Discussion by Spiritreader** [AutoDarkMode/Windows-Auto-Night-Mode#593] [Q&A]
# Discussion / Q&A Thread

If you have any questions about this release feel free to ask here!

# Changelog

This release brings many more theme switching fixes for Windows 11 22H2 and adds many new QoL features. It also brings back the known and tested switching engine for versions older than Windows 10 21H2, as they had trouble functioning correctly with the change needed for Windows 11.
In addition, there are countless under the hood fixes that should increase the theme switching reliability compared to 10.2.
We've also reduced the number of occasions where UI elements don't update properly. However, some remain because theme switching in 22H2 is quite buggy by itself.

All in all, this is the biggest update since the 10.0 release, we hope you like it!

## Known limitations
Auto Dark Mode comes with a few limitations, either due to Windows or because of technical reasons. 
You can find the full list here: [List of known limitations](https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/wiki/Known-limitations)

# Changelog (10.3.0.90)

## New Features
- Use Windows Night Light as source for your light and dark switch times. You can of course still configure an offset in ADM (#489, @thebigsmileXD)!
- Pause automatic theme switching once (postpones until next sunrise / sunset time dynamically) (#439).
- Pause automatic theme switching for a short amount of time (#204, #439).
- Managed theme switch postponing via the UI on the time page:
  - Pause and resume theme switching via the UI on the Time page.
  - See which modules are holding up a theme switch and for how long (if they have an expiry time).
- Show a notification and delay or postpone this action interactively: (#439, #204, #441).
  - Before an automatic theme switch is applied and the theme needs to be updated.
  - When auto dark mode is started.
- New hotkeys:
  - Turn on/off automatic theme switching.
  - Pause theme switching until next sunrise/sunset.
  - Toggle themes between light and dark using a single hotkey. If you switch to a theme that doesn't match the time, we will auto-pause theme switching and notify you. 
- New context menu options in the notification area / tray that provide the same functionality as the hotkeys.
- Toggle colorization for title bars and window borders for Windows 10 (does work on Windows 11 for some windows, but not all).
- Ignore certain options like background, cursors and sounds when using the Windows theme mode (#103, #492, @namazso).

## Improvements
- Uses an additional, more modern theme switching method that supports more options and makes ADM more reliable (Huge thanks to @namazso for the awesome contribution)!
- Proper support for in-lockscreen theme switching. Windows 11 will switch as soon as you unlock your PC, Windows 10 will switch themes during the lockscreen.
- Clicking on the tray icon when the app is minimized will now instantly bring it to the front.
- Many stability improvements for theme state synchronization that include but aren't limited to:
  - More reliable wallpaper retrieval from an existing custom theme (for support with third party wallpaper tools),
  - Better detection when an unsaved Windows theme is used.
  - Improve potential wallpaper flickering by reducing the number of required synchronization calls.
  - More robust detection of the currently active windows theme.
- Single monitor and multi monitor wallpapers as well as solid color are now being retrieved when the wallpaper picker is enabled and when Auto Dark Mode starts for a smoother wallpaper switching experience with our own picker.
- The context menu in the tray now supports dark mode.
- The wallpaper page now requests a theme refresh when the setting is toggled on, which results in a slightly better user experience.
- We redesigned the personalization page to look a bit more modern.
- Added information text when pressing the copy button on the about page.
- Notification area icon and hover text shows if auto theme switching is paused or disabled (#560)

## Fixes
- Fixed an issue where ADM stopped working with builds older than 19044 (#508).
- Fixed an issue on Windows 10 that caused multi users to experience missed theme switches on a user account when they were currently logged into another. You can toggle support for this on or off in the settings menu (#544).
- Fixed an issue where the Windows default themes were not correctly synchronized.
- Fixed an issue where custom theme synchronization was dependent on the system language and would only work correctly in English.
- Fixed an issue where activated slideshows in the source theme would cause the wallpaper picker to fail setting wallpapers,
- Fixed an issue where solid color backgrounds failed to synchronize (#547).
- Fixed multiple issues where incorrect theme paths were being retrieved, causing either too much or too little theme synchronizations and applications.
- Fixed an issue on Windows 10 where the "Restore themes when changed externally" setting would cause excessive theme updates by rate limiting the updates.
- Fixed an issue where the updater would parse whitelisted files with case-sensitivity.
- Fixed an issue where the theme picker page was always forcing a theme refresh when it loads, even when the setting is disabled.

## Changes
- ADM now makes an unmanaged copy of every theme when windows theme mode is enabled. This improves theme switching reliability on Windows 11 22H2 and should not affect the experience for users of other versions.
- Moved Help button to the bottom navbar entries (#557, #549, thanks @PtrStruct)
- Allow modules to run before or after theme synchronization.
- New SwitchEvent: Api. It is meant for calls that shouldn't invoke user notifications. The event is currently used when themes are set via the shell.
- The update button can now be pressed multiple times.
- System resume events can **no longer be turned off**. They are mandatory for ensuring that theme switching is working at all times.
- Use legacy theme switching for builds older Win10 21H1
- UI uses Mica for 22H2 and newer.

## Translation Updates
- NEW: Greek translation (@MariosMpalas).
- Update Japanese translation (@maboroshin).
- Update Turkish translation (@cmhrky).
- Update Spanish translation (@DanielSevillano)
- Update Bahasa Indonesia translation (@pc-v2)
- Update Vietnamese translation (@KhoiCanDev)
- Update Spanish translation (@DanielSevillano, @alealclag).
- Update Italian translation (@albertopasqualetto).
- Update Ukrainian translation (@MykhaiIo).
- Update Czech translation (@drcryo)
- Wording & typo fixes (@tooomm)

<hr /><em>This discussion was created from the release <a href='https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/releases/tag/10.3.0.90'>Auto Dark Mode X 10.3.0</a>.</em>

**Comment by HanburuSiPairotto77** [comment]
Hey quick question, I was wondering as to why ADM was making an unmanaged copy of every theme, but it is explained here, my question is, what does it mean by "unmanaged"?

**Reply by Spiritreader** [reply]
I'll quickly lay out the differences between managed and unmanaged

## Managed

By default ADM will synchronize theme settings from Windows.
Whenever the active Windows theme changes, the internal (or managed) representation of that theme file will update accordingly, We achieve that by parsing the ini file and deserializing the theme into a class that we built. When appliyng the theme, we will *serialize* that class back to an .ini file.

https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/blob/c1effc3643a08aaf1a28a882bf035e38ae36dcb0/AutoDarkModeSvc/Handlers/ThemeFiles/ThemeFile.cs#L198

https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/blob/c1effc3643a08aaf1a28a882bf035e38ae36dcb0/AutoDarkModeSvc/Handlers/ThemeFiles/ThemeFile.cs#L274

Essentially, the supported features are entirely dependent on whether we implement that functionality into our theme file class.

https://github.com/AutoDarkMode/Windows-Auto-Night-Mode/blob/master/AutoDarkModeSvc/Handlers/ThemeFiles/ThemeFileMappedSections.cs

## Unmanaged

If you use the windows theme mode, however, ADM will create a copy of the respective light and dark theme you have selected.
It then adds a bit of metadata so that we can keep track of which theme is supposed to be active at the moment.
This, by the way, is why we need to create a copy, because some Windows default themes are read-only.
The required metadata is also the reason why there are two files: An ADMUnmanagedLight and ADMUnmanagedDark theme file. 

Now, the reason theme mode is also called **unmanaged mode** is that individual settings within the theme file will not pass through our internal theme file representation (or: managed by Auto Dark Mode) when saving them before we apply that theme.
As such, we don't re-serialize that theme file based on our own representation, but instead only process it as an array of lines. This has the advantage that we don't need to explicitly support all features in a theme file. For people that want to use functionality that is not exposed in managed mode, they have the option to do so with that.
